ls1intum/Artemis
# Fix Deletion Test Runs with Missing Participations

### Checklist

### Motivation and Context
 This solves the problem that test runs cannot be deleted if a participation is missing (deleted separately in some way, e.g. Editor in Editor and the Assignment Repository there). See also https://sentry.io/organizations/ls1intum/issues/2269187209/?environment=prod&project=1440029&query=is:unresolved

### Description
- Added test for a test run missing a participation being deleted
- Add check before using `iterator().next()` on the participation
- Refactor the code a bit to remove redundant logic (this comes at no cost since the lists are small, and `otherTestRunsOfInstructor` and `allInstructorTestRunExercises` will just be empty)

### Steps for Testing
0. Review the code
1. Create an exam with a programming exercise and another programming exercise
2. Do a test run with participating in both exercises
3. Do a second test run with the same instructor
4. Go to the programming exercise in the exercise groups view and go on "Edit in Editor"
5. (Create and then) delete the assignment repository
6. Delete one test run
7. Check that the participation in the other exercise is still present
    (I recommend checking for present participations by going to the project repositories overview in Bitbucket)
8. Delete the other test run
9. Check that all participations are gone

### Test Coverage
Should be the same or better (due to redundant code being removed and the new test case)
